File: frontend-web/src/app/services/carrito.service.ts
import { Injectable, inject, signal } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable, tap } from 'rxjs';
import { environment } from '../../environments/environment';
import { AuthService } from './auth.service';

@Injectable({
  providedIn: 'root'
})
export class CarritoService {
  private http = inject(HttpClient);
  private authService = inject(AuthService);
  private apiUrl = `${environment.apiUrl}/carrito`;

  public carritoCount = signal<number>(0);

  private get headers(): HttpHeaders {
    return new HttpHeaders({
      'Content-Type': 'application/json',
      Authorization: `Bearer ${this.authService.token}`
    });
  }

  getCarrito(): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/`, { headers: this.headers }).pipe(
      tap(c => this.carritoCount.set(c?.total_items || 0))
    );
  }

  agregarItem(varianteId: number, cantidad: number = 1): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/agregar`, { variante_id: varianteId, cantidad }, { headers: this.headers }).pipe(
      tap(c => this.carritoCount.set(c?.total_items || 0))
    );
  }

  actualizarItem(itemId: number, cantidad: number): Observable<any> {
    return this.http.put<any>(`${this.apiUrl}/item/${itemId}`, { cantidad }, { headers: this.headers }).pipe(
      tap(c => this.carritoCount.set(c?.total_items || 0))
    );
  }

  eliminarItem(itemId: number): Observable<any> {
    return this.http.delete<any>(`${this.apiUrl}/item/${itemId}`, { headers: this.headers }).pipe(
      tap(c => this.carritoCount.set(c?.total_items || 0))
    );
  }

  vaciarCarrito(): Observable<any> {
    return this.http.delete<any>(`${this.apiUrl}/vaciar`, { headers: this.headers }).pipe(
      tap(() => this.carritoCount.set(0))
    );
  }

  refreshCount(): void {
    if (this.authService.isLoggedIn) {
      this.getCarrito().subscribe({
        next: () => {},
        error: () => {}
      });
    }
  }
}

File: frontend_mobile/lib/services/carrito_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import '../constants.dart';

class CarritoService {
  Future<String?> _getToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('jwt_token');
  }

  Future<Map<String, dynamic>?> getCarrito() async {
    try {
      final token = await _getToken();
      final response = await http.get(
        Uri.parse('${Constants.apiUrl}/carrito/'),
        headers: {
          'Accept': 'application/json',
          if (token != null) 'Authorization': 'Bearer $token',
        }
      );
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      }
      return null;
    } catch (_) { return null; }
  }

  Future<Map<String, dynamic>?> agregarItem(int varianteId, int cantidad) async {
    try {
      final token = await _getToken();
      final response = await http.post(
        Uri.parse('${Constants.apiUrl}/carrito/agregar'),
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json',
          if (token != null) 'Authorization': 'Bearer $token',
        },
        body: jsonEncode({
          'variante_id': varianteId,
          'cantidad': cantidad,
        }),
      );
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      }
      return null;
    } catch (_) { return null; }
  }

  Future<Map<String, dynamic>?> actualizarItem(int itemId, int cantidad) async {
    try {
      final token = await _getToken();
      final response = await http.put(
        Uri.parse('${Constants.apiUrl}/carrito/item/$itemId'),
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json',
          if (token != null) 'Authorization': 'Bearer $token',
        },
        body: jsonEncode({'cantidad': cantidad}),
      );
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      }
      return null;
    } catch (_) { return null; }
  }

  Future<Map<String, dynamic>?> eliminarItem(int itemId) async {
    try {
      final token = await _getToken();
      final response = await http.delete(
        Uri.parse('${Constants.apiUrl}/carrito/item/$itemId'),
        headers: {
          'Accept': 'application/json',
          if (token != null) 'Authorization': 'Bearer $token',
        }
      );
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      }
      return null;
    } catch (_) { return null; }
  }
}
